format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressor.java

```diff
@@ -0,0 +1,266 @@
+
+  public FixedLengthCharsetTransformingDecompressor(FixedLengthCharset sourceEncoding) {
+    this.sourceEncoding = sourceEncoding;
+    this.decoder = sourceEncoding.getCharset().newDecoder();
+    this.encoder = targetCharset.newEncoder();
+  }
+
+  @Override
+  public void setInput(byte[] b, int off, int len) {
+    //Expand incoming buffer if needed.
+    if (inputByteBuffer.remaining() < len) {
+      //Allocate new buffer that can fill the existing input + newly received bytes
+      ByteBuffer newIncomingBuffer = ByteBuffer.allocate(len + inputByteBuffer.capacity());
+
+      //Set up incoming buffer for reads and copy contents into new buffer.
+      inputByteBuffer.flip();
+      newIncomingBuffer.put(inputByteBuffer);
+
+      inputByteBuffer = newIncomingBuffer;
+    }
+
+    //Copy incoming payload into Input Byte Buffer
+    inputByteBuffer.put(b, off, len);
+    inputByteBuffer.flip();
+
+    //Set up char buffer for writes
+    decodedCharBuffer.compact();
+
+    //Expand the char buffer if needed.
+    if (decodedCharBuffer.capacity() < inputByteBuffer.limit() / sourceEncoding.getNumBytesPerCharacter()) {
+      decodedCharBuffer = CharBuffer.allocate(inputByteBuffer.limit() / sourceEncoding.getNumBytesPerCharacter());
+    }
+
+    //Decode bytes from the input buffer into the Decoded Char Buffer
+    decodeByteBufferIntoCharBuffer(inputByteBuffer);
+
+    //Set up decoded char buffer for reads.
+    decodedCharBuffer.flip();
+
+    //Set up incoming buffer for writes.
+    inputByteBuffer.compact();
+
+  }
+
+  /**
+   * Note that we only ask for additional input once we have completely depleted out outgoing buffer.
+   */
+  @Override
+  public boolean needsInput() {
+    return decodedCharBuffer.remaining() == 0;
+  }
+
+  @Override
+  public void setDictionary(byte[] b, int off, int len) {
+    //no-op
+  }
+
+  @Override
+  public boolean needsDictionary() {
+    return false;
+  }
+
+  @Override
+  public boolean finished() {
+    return decodedCharBuffer.remaining() == 0 && partialOutputByteBuffer.remaining() == 0;
+  }
+
+  @Override
+  public int decompress(byte[] b, int off, int len) throws IOException {
+
+    //Allocate new outgoing buffer
+    ByteBuffer encodedBuffer = ByteBuffer.allocate(len - off);
+
+    //Consume any remaining bytes from a previous decompress invocation.
+    while (partialOutputByteBuffer != null && partialOutputByteBuffer.hasRemaining() && encodedBuffer.hasRemaining()) {
+      encodedBuffer.put(partialOutputByteBuffer.get());
+    }
+
+    //Encode as many characters as possible into the Encoded Buffer.
+    encodeCharBufferIntoByteBuffer(encodedBuffer);
+
+    // Handle the case where the outgoing buffer can still fit additional space.
+    // This means we need to encode one extra character and add as many bytes as possible into the output buffer.
+    if (decodedCharBuffer.remaining() > 0 && encodedBuffer.remaining() > 0) {
+      encodePartialCharacter(encodedBuffer);
```

Comment:
This is a pretty heavy operation. I think we need it only when we could not encode anything. It's totally fine to fil less than `len` as soon as we fill at least something. We are not required to fill each block exactly to the length.

Reply:
Good catch. I've added additional logic to only execute this operation when there is no other option (Meaning, we were not able to add a single byte to the output.

Reply:
I actually think this can be our problem and not partition alignment. If we read something from input and then don't return it to user we can't give proper position information and records are lost